### ISSUE TITLE
Adapters are not setup for non-Flux functions in non-catalog environments

### ISSUE THREAD

**Issue by markusgulden** [spring-cloud/spring-cloud-function#252]
_Basically the description below is valid for all adapters. The only thing I'd like to add is that this happens in the scenario where FunctionCatalog is not available and functions are retrieved directly from BeanFactory_

====

Unfortunately, I didn't get any answer on my [question on StackOverflow](https://stackoverflow.com/questions/54356212/spring-cloud-function-expects-fluxstring-instead-of-string-when-deploying-on) so far, so maybe somebody can help me here.

In the past, I've implemented a function and deployed it on AWS Lambda, which worked perfectly.
But when I now try deploy the same function on Azure Functions, that does not work:

`24.01.19 21:58:50] Caused by: java.lang.ClassCastException: reactor.core.publisher.FluxJust cannot be cast to java.lang.String
[24.01.19 21:58:50]     at de.margul.awstutorials.springcloudfunction.function.EchoFunction.apply(EchoFunction.java:9)
[24.01.19 21:58:50]     at org.springframework.cloud.function.adapter.azure.AzureSpringBootRequestHandler.handleRequest(AzureSpringBootRequestHandler.java:56)
[24.01.19 21:58:50]     at de.margul.awstutorials.springcloudfunction.azure.handler.FunctionHandler.execute(FunctionHandler.java:19)
[24.01.19 21:58:50]     ... 16 more `

I've created a [minimal example](https://github.com/markusgulden/playground/tree/master/springcloudfunction.example) to reproduce the problem.
You have to install the [Azure Function Core Tools](https://docs.microsoft.com/en-us/azure/azure-functions/functions-run-local) to run the example locally.
After installation, you can build and run the function using:
`mvn package`
`mvn azure-functions:run`

To invoke the function, you have to call:
`curl -d 'test' -X POST http://localhost:7071/api/echoFunction`

Note: you can switch between the working and the not working function by changing the _function.name_ attribute in the _local.settings.json_.

This function works:
```
@Component
public class FluxEchoFunction implements Function<Flux<String>, Flux<String>> {
    @Override
    public Flux<String> apply(Flux<String> m) {
        return Flux.just("Echo: " + m);
    }
}
```

and this function does not work (throws the exception above):

```
@Component
public class PlainEchoFunction implements Function<String, String> {

    @Override
    public String apply(String m) {
        return "Echo: " + m;
    }
}
```

I think this is the relevant part [in the documentation](https://cloud.spring.io/spring-cloud-static/spring-cloud-function/2.0.0.RELEASE/single/spring-cloud-function.html#_function_catalog_and_flexible_function_signatures):

> One of the main features of Spring Cloud Function is to adapt and support a range of type signatures for user-defined functions, while providing a consistent execution model. That’s why all user defined functions are transformed into a canonical representation by FunctionCatalog, using primitives defined by the Project Reactor (i.e., Flux and Mono). Users can supply a bean of type Function, for instance, and the FunctionCatalog will wrap it into a Function,Flux>.

However, according to that, I would expect that I don't have to define my function like `Flux<String> apply(Flux<String> m)`, right?

**Comment by markusgulden** [comment]
@olegz , do you have any idea whether this is a bug, or am I just doing something wrong?

**Comment by olegz** [comment]
@markusgulden could you please give it another shot with the latest snapshot? I'll do my due diligence with your sample, but I believe the latest commit may have indirectly fixed this issue.

Anyway, let me know if/when you get a chance

**Comment by markusgulden** [comment]
@olegz , unfortunately, die latest snapshot does not seem to have the problem fixed.

However, I found another way to solve the issue.

Until now, my function handler looked like this:
```
public class FunctionHander extends AzureSpringBootRequestHandler<String, String> {
    @FunctionName("echoFunction")
    public String execute(@HttpTrigger(name = "req", methods = {
            HttpMethod.POST }, authLevel = AuthorizationLevel.ANONYMOUS) HttpRequestMessage<String> entity,
            ExecutionContext context) {
        return handleRequest(entity.getBody(), context);
    }
    @Bean
    public FluxEchoFunction fluxEchoFunction() {
        return new FluxEchoFunction();
    }

    @Bean
    public PlainEchoFunction plainEchoFunction() {
        return new PlainEchoFunction();
    }
}
```

When I change it now to the following, I can call my PlainEchoFunction without errors:
```
@SpringBootConfiguration
public class FunctionHander extends AzureSpringBootRequestHandler<String, String> implements ApplicationContextInitializer<GenericApplicationContext>{
    @FunctionName("echoFunction")
    public String execute(@HttpTrigger(name = "req", methods = {
            HttpMethod.POST }, authLevel = AuthorizationLevel.ANONYMOUS) HttpRequestMessage<String> entity,
            ExecutionContext context) {
        return handleRequest(entity.getBody(), context);
    }

    @Override
    public void initialize(GenericApplicationContext context) {
      context.registerBean("plainEchoFunction", FunctionRegistration.class,
          () -> new FunctionRegistration<>(new PlainEchoFunction())
              .type(FunctionType.from(String.class).to(String.class)));
    }
}
```

Does that help with analyzing the issue?

**Comment by olegz** [comment]
It certainly does, let me dig in

**Comment by olegz** [comment]
@markusgulden As you may have seen I've commented on your [SO question](https://stackoverflow.com/questions/54356212/spring-cloud-function-expects-fluxstring-instead-of-string-when-deploying-on/54822230#54822230). Basically we have a bug. Will take care of it before next release

**Comment by olegz** [comment]
@markusgulden your input definitely helped. I figured out the issue and it is actually affecting all adapters. So, I'll push by Monday after I move some things around

**Comment by markusgulden** [comment]
Glad to hear that you found the bug, @olegz 
Because even if I made it to get the simple example from above working, the issue still occurs with something more complex like a Function<Message<I>, Message<O>>.

**Comment by olegz** [comment]
So, the issue is really not related to Azure and rather common across multiple adapters, so I'll modify the title slightly. 
The issue has been fixed indirectly on the master with [this commit](https://github.com/spring-cloud/spring-cloud-function/commit/0189c578efe085e273eece62f54bf8bdb86948f4).
I am about to merge backport to 2.0.x after which I'll close it, so give it a shot as I assume there may be some additional issues lingering

**Comment by olegz** [comment]
resolved with https://github.com/spring-cloud/spring-cloud-function/commit/c20c87d1dd0ea31ea377bc5c2bba398d149c5016 and https://github.com/spring-cloud/spring-cloud-function/commit/0189c578efe085e273eece62f54bf8bdb86948f4

**Comment by ruewan** [comment]
I get the same thing with spring-cloud-function-adapter-aws 2.0.1.RELEASE. If I Use @Import to load my application config instead of @SpringBootApplication I get the following error.
java.lang.ClassCastException: reactor.core.publisher.FluxJust cannot be cast to com.amazonaws.services.lambda.runtime.events.DynamodbEvent
	at com.example.pet.function.HandleDBEvent.apply(HandleDBEvent.java:20)
	at org.springframework.cloud.function.adapter.aws.SpringFunctionInitializer.apply(SpringFunctionInitializer.java:132)
	at org.springframework.cloud.function.adapter.aws.SpringBootRequestHandler.handleRequest(SpringBootRequestHandler.java:48)

I am not sure what extra configuration I need if I am not using the autoconfig.

**Comment by olegz** [comment]
@ruewan What made you believe it would work in 2.0.1.RELEASE when it was reported as a bug? 
It was fixed and is available in 2.1.0.M1 as well as the snapshots
